bigtable/client/internal/strong_type.h

```diff
@@ -0,0 +1,45 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_INTERNAL_STRONG_TYPE_H
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_INTERNAL_STRONG_TYPE_H
+
+#include "bigtable/client/version.h"
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+namespace internal {
+
+/**
+ * Helper class to avoid cases where user function accepts parameter of same
+ * primitive type
+ * @tparam T primitive type.
+ * @tparam Parameter Name of the parameter.
```

Comment:
Consider:
```C++
/**
 * A simple implementation of the strong type C++ idiom.
 *
 * In many cases the same class or type (`int`, `double`, `std::string`) is used to represent very different things.
 * It is desirable to create a strong type that wraps the basic type to avoid common mistakes in an API, such as
 * passing parameters in the wrong order.  For example, consider a function that takes a distance and speed 
 * parameters:
 * @code
 *  double f(double speed, double distance);
 * @endcode
 *
 * This function can used (by mistake), as:
 * @code
 * double my_distance = ..;
 * double my_speed = ...;
 * double x = f(my_distance, my_speed);
 * @endcode
 *
 * One can avoid those mistakes by using a `StrongType`:
 * @code
 * using Distance = StrongType<double, struct DistanceTag>;
 * using Speed = StrongType<double, struct SpeedTag>;
 * double f(Speed speed, Distance distance);
 * @endcode
 *
 * @tparam T the type wrapped by the strong type.
 * @tparam Parameter a formal parameter to create different instantiations of `StrongType<T, ...>`.
 */
```